
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Ag1000G phase 3 SNP data release - cloud data access guide &#8212; MalariaGEN vector open data guide</title>
    
  <link rel="stylesheet" href="../_static/css/index.73d71520a4ca3b99cfee5594769eaaae.css">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/myfile.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.3da636dd464baa7582d2.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Ag1000G phase 3 SNP data release - data download guide" href="download.html" />
    <link rel="prev" title="Ag1000G phase 3 - summary of samples" href="intro.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../index.html">
  
  <img src="../_static/logo.png" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">MalariaGEN vector open data guide</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../landing-page.html">
   MalariaGEN vector open data guide
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Ag1000G phase 3 (Ag3)
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="intro.html">
   Ag1000G phase 3 - summary of samples
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Ag1000G phase 3 SNP data release - cloud data access guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="download.html">
   Ag1000G phase 3 SNP data release - data download guide
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/ag3-snp/cloud.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/executablebooks/jupyter-book/master?urlpath=tree/ag3-snp/cloud.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#about-this-guide">
   About this guide
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-hosting">
   Data hosting
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#sample-sets">
   Sample sets
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#sample-metadata">
   Sample metadata
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snp-sites-and-alleles">
   SNP sites and alleles
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#site-filters">
   Site filters
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snp-genotypes">
   SNP genotypes
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#further-information">
   Further information
  </a>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="ag1000g-phase-3-snp-data-release-cloud-data-access-guide">
<h1>Ag1000G phase 3 SNP data release - cloud data access guide<a class="headerlink" href="#ag1000g-phase-3-snp-data-release-cloud-data-access-guide" title="Permalink to this headline">¶</a></h1>
<p><strong>14 January 2021</strong></p>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This notebook provides information about how to access data from the <a class="reference external" href="https://www.malariagen.net/data/ag1000g-phase3-snp">MalariaGEN Anopheles gambiae 1000 Genomes project (Ag1000G) phase 3 SNP data release</a> via Google Cloud. This includes sample metadata and single nucleotide polymorphism (SNP) calls.</p>
<p>Ag1000G phase 3 includes SNP calls for 3,081 individual mosquitoes, including 2,784 mosquitoes collected from natural populations in 19 countries. The map below provides an overview of the collection locations and the numbers of samples broken down by species. Further information about the samples and the studies which contributed them is available from the <a class="reference external" href="https://storage.googleapis.com/vo_agam_release/v3/ag1000g-phase3-contributing-studies.pdf">contributing studies document</a>.</p>
<img src="ag3-map.png" alt="map" width="600"/>
<p>Please note that data from Ag1000G phase 3 are being released prior to publication. You are free to use these data for any purpose, but the data are under a publication embargo described further in the <a class="reference external" href="https://www.malariagen.net/data/terms-use/ag1000g-terms-use">Ag1000G terms of use</a>. If you have any questions about the terms of use, please email data&#64;malariagen.net.</p>
<p>If you have any questions about this guide or how to use the data, please <a class="reference external" href="https://github.com/malariagen/vector-public-data/discussions/new">start a new discussion</a> on the malariagen/vector-public-data repo on GitHub. If you find any bugs, please <a class="reference external" href="https://github.com/malariagen/vector-public-data/issues/new/choose">raise an issue</a>.</p>
</div>
<div class="section" id="about-this-guide">
<h2>About this guide<a class="headerlink" href="#about-this-guide" title="Permalink to this headline">¶</a></h2>
<p>This guide is written as a Jupyter notebook. It illustrates how to read data directly from the cloud, without having to first download any data locally. This notebook can be run from any computer, but will work best when run from a compute node within Google Cloud, because it will be physically closer to the data and so data transfer is faster. For example, this notebook can be run via <a class="reference external" href="https://colab.research.google.com/">Google Colab</a> or <a class="reference external" href="https://cloud.google.com/datalab/docs">Google Cloud Datalab</a> which are interactive computing service running in the cloud. Here is a link to <a class="reference external" href="https://colab.research.google.com/github/malariagen/vector-public-data/blob/master/docs/ag3-snp-download-guide.ipynb">open this notebook in Google Colab</a>.</p>
<p>This guide covers accessing SNP calls and associated sample metadata only. If you wish to access raw sequence reads or sequence read alignments you will need to download data locally, see the <span class="xref myst">Ag1000G phase 3 SNP download guide &#64;&#64;TODO link</span>.</p>
</div>
<div class="section" id="data-hosting">
<h2>Data hosting<a class="headerlink" href="#data-hosting" title="Permalink to this headline">¶</a></h2>
<p>All data required for this notebook is hosted on Google Cloud Storage (GCS). Sample metadata in CSV format and SNP calls in Zarr format are hosted on GCS in the <code class="docutils literal notranslate"><span class="pre">vo_agam_release</span></code> bucket, which is a multi-region bucket located in the United States. All data hosted in GCS are publicly accessible and do not require any authentication to access.</p>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h2>
<p>Running this notebook requires the following Python packages to be installed: numpy, dask, zarr, gcsfs, fsspec, scikit-allel.</p>
<p>These packages can be installed via pip or conda. E.g.:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip install -q numpy dask<span class="o">[</span>array<span class="o">]</span> zarr gcsfs fsspec scikit-allel
</pre></div>
</div>
</div>
</div>
<p>To make accessing these data more convenient, we’ve also created a <a class="reference external" href="https://github.com/malariagen/malariagen-data-python">malariagen_data Python package</a>, which is available from PyPI. This is experimental so please let us know if you find any bugs or have any suggestions. The <code class="docutils literal notranslate"><span class="pre">malariagen_data</span></code> package can be installed via pip, e.g.:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip install -q malariagen-data
</pre></div>
</div>
</div>
</div>
<p>Once installed, data access via GCS is set up via the following code:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">malariagen_data</span>
<span class="n">ag3</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Ag3</span><span class="p">(</span><span class="s2">&quot;gs://vo_agam_release/&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="sample-sets">
<h2>Sample sets<a class="headerlink" href="#sample-sets" title="Permalink to this headline">¶</a></h2>
<p>Data in this release are organised into 28 sample sets. Each of these sample sets corresponds to a set of mosquito specimens contributed by a collaborating study. Depending on your objectives, you may want to access data from only specific sample sets, or all sample sets.</p>
<p>To see which sample sets are available, load the sample set manifest into a <a class="reference external" href="https://pandas.pydata.org/">pandas</a> dataframe:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_sample_sets</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">sample_sets</span><span class="p">()</span>
<span class="n">df_sample_sets</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_set</th>
      <th>sample_count</th>
      <th>release</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AG1000G-AO</td>
      <td>81</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AG1000G-BF-A</td>
      <td>181</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AG1000G-BF-B</td>
      <td>102</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AG1000G-BF-C</td>
      <td>13</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AG1000G-CD</td>
      <td>76</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>5</th>
      <td>AG1000G-CF</td>
      <td>73</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>6</th>
      <td>AG1000G-CI</td>
      <td>80</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>7</th>
      <td>AG1000G-CM-A</td>
      <td>303</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>8</th>
      <td>AG1000G-CM-B</td>
      <td>97</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>9</th>
      <td>AG1000G-CM-C</td>
      <td>44</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>10</th>
      <td>AG1000G-FR</td>
      <td>23</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>11</th>
      <td>AG1000G-GA-A</td>
      <td>69</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>12</th>
      <td>AG1000G-GH</td>
      <td>100</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>13</th>
      <td>AG1000G-GM-A</td>
      <td>74</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>14</th>
      <td>AG1000G-GM-B</td>
      <td>31</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>15</th>
      <td>AG1000G-GM-C</td>
      <td>174</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>16</th>
      <td>AG1000G-GN-A</td>
      <td>45</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>17</th>
      <td>AG1000G-GN-B</td>
      <td>185</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>18</th>
      <td>AG1000G-GQ</td>
      <td>10</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>19</th>
      <td>AG1000G-GW</td>
      <td>101</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>20</th>
      <td>AG1000G-KE</td>
      <td>86</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>21</th>
      <td>AG1000G-ML-A</td>
      <td>60</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>22</th>
      <td>AG1000G-ML-B</td>
      <td>71</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>23</th>
      <td>AG1000G-MW</td>
      <td>41</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>24</th>
      <td>AG1000G-MZ</td>
      <td>74</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>25</th>
      <td>AG1000G-TZ</td>
      <td>300</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>26</th>
      <td>AG1000G-UG</td>
      <td>290</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>27</th>
      <td>AG1000G-X</td>
      <td>297</td>
      <td>v3</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>The sample set identifiers all start with “AG1000G-” followed by the two-letter code of the country from which samples were collected (e.g., “AO” is Angola). Where there are multiple sample sets from the same country, these have been given alphabetical suffixes, e.g., “AG1000G-BF-A”, “AG1000G-BF-B” and “AG1000G-BF-C” are three sample sets from Burkina Faso.</p>
<p>These country code suffixes are just a convenience to help remember which sample sets contain which data, please see the sample metadata for more precise location information. Please note also that sample set AG1000G-GN-B contains samples from both Guinea and Mali.</p>
</div>
<div class="section" id="sample-metadata">
<h2>Sample metadata<a class="headerlink" href="#sample-metadata" title="Permalink to this headline">¶</a></h2>
<p>Data about the samples that were sequenced to generate this data resource are available, including the time and place of collection, the gender of the specimen, and our call regarding the species of the specimen. These are organised by sample set.</p>
<p>E.g., load sample metadata for the AG1000G-BF-A and AG1000G-BF-B sample sets:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_samples</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">(</span><span class="n">cohort</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AG1000G-BF-A&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-B&quot;</span><span class="p">])</span>
<span class="n">df_samples</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_id</th>
      <th>partner_sample_id</th>
      <th>contributor</th>
      <th>country</th>
      <th>location</th>
      <th>year</th>
      <th>month</th>
      <th>latitude</th>
      <th>longitude</th>
      <th>sex_call</th>
      <th>sample_set</th>
      <th>release</th>
      <th>aim_fraction_colu</th>
      <th>aim_fraction_arab</th>
      <th>species_gambcolu_arabiensis</th>
      <th>species_gambiae_coluzzii</th>
      <th>species</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AB0085-Cx</td>
      <td>BF2-4</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Pala</td>
      <td>2012</td>
      <td>7</td>
      <td>11.150</td>
      <td>-4.235</td>
      <td>F</td>
      <td>AG1000G-BF-A</td>
      <td>v3</td>
      <td>0.024</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AB0086-Cx</td>
      <td>BF2-6</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Pala</td>
      <td>2012</td>
      <td>7</td>
      <td>11.150</td>
      <td>-4.235</td>
      <td>F</td>
      <td>AG1000G-BF-A</td>
      <td>v3</td>
      <td>0.038</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AB0087-C</td>
      <td>BF3-3</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Bana</td>
      <td>2012</td>
      <td>7</td>
      <td>11.233</td>
      <td>-4.472</td>
      <td>F</td>
      <td>AG1000G-BF-A</td>
      <td>v3</td>
      <td>0.982</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AB0088-C</td>
      <td>BF3-5</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Bana</td>
      <td>2012</td>
      <td>7</td>
      <td>11.233</td>
      <td>-4.472</td>
      <td>F</td>
      <td>AG1000G-BF-A</td>
      <td>v3</td>
      <td>0.990</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AB0089-Cx</td>
      <td>BF3-8</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Bana</td>
      <td>2012</td>
      <td>7</td>
      <td>11.233</td>
      <td>-4.472</td>
      <td>F</td>
      <td>AG1000G-BF-A</td>
      <td>v3</td>
      <td>0.975</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>278</th>
      <td>AB0533-C</td>
      <td>BF13-18</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Souroukoudinga</td>
      <td>2014</td>
      <td>7</td>
      <td>11.235</td>
      <td>-4.535</td>
      <td>F</td>
      <td>AG1000G-BF-B</td>
      <td>v3</td>
      <td>0.021</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>279</th>
      <td>AB0536-C</td>
      <td>BF13-31</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Souroukoudinga</td>
      <td>2014</td>
      <td>7</td>
      <td>11.235</td>
      <td>-4.535</td>
      <td>F</td>
      <td>AG1000G-BF-B</td>
      <td>v3</td>
      <td>0.025</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>280</th>
      <td>AB0537-C</td>
      <td>BF13-32</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Souroukoudinga</td>
      <td>2014</td>
      <td>7</td>
      <td>11.235</td>
      <td>-4.535</td>
      <td>F</td>
      <td>AG1000G-BF-B</td>
      <td>v3</td>
      <td>0.029</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>281</th>
      <td>AB0538-C</td>
      <td>BF13-33</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Souroukoudinga</td>
      <td>2014</td>
      <td>7</td>
      <td>11.235</td>
      <td>-4.535</td>
      <td>F</td>
      <td>AG1000G-BF-B</td>
      <td>v3</td>
      <td>0.018</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>282</th>
      <td>AB0408-C</td>
      <td>BF14-20</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Bana</td>
      <td>2014</td>
      <td>7</td>
      <td>11.233</td>
      <td>-4.472</td>
      <td>F</td>
      <td>AG1000G-BF-B</td>
      <td>v3</td>
      <td>0.986</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
  </tbody>
</table>
<p>283 rows × 17 columns</p>
</div></div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">sample_id</span></code> column gives the sample identifier used throughout all Ag1000G analyses.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">country</span></code>, <code class="docutils literal notranslate"><span class="pre">location</span></code>, <code class="docutils literal notranslate"><span class="pre">latitude</span></code> and <code class="docutils literal notranslate"><span class="pre">longitude</span></code> columns give the location where the specimen was collected.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">year</span></code> and <code class="docutils literal notranslate"><span class="pre">month</span></code> columns give the approximate date when the specimen was collected.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">sex_call</span></code> column gives the gender as determined from the sequence data.</p>
<p>Pandas can be used to explore and query the sample metadata in various ways. E.g., here is a summary of the numbers of samples by sample set, collection location, year and species:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_samples</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;sample_set&quot;</span><span class="p">,</span> <span class="s2">&quot;country&quot;</span><span class="p">,</span> <span class="s2">&quot;year&quot;</span><span class="p">,</span> <span class="s2">&quot;species&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>sample_set    country       year  species                      
AG1000G-BF-A  Burkina Faso  2012  coluzzii                         82
                                  gambiae                          98
                                  intermediate_gambiae_coluzzii     1
AG1000G-BF-B  Burkina Faso  2014  arabiensis                        3
                                  coluzzii                         53
                                  gambiae                          46
dtype: int64
</pre></div>
</div>
</div>
</div>
<p>Note that samples within a sample set may belong to different species. For convenience, we have made a species call for all samples in Ag1000G phase 3, using the genomic data. Calling species is not always straightforward, and we have used two different methods for species calling, ancestry informative markers (AIM) and principal components analysis (PCA). When loading the sample metadata, the AIM species calls are included by default. The results of these two different methods generally agree, although there are some populations where results are different, particularly in Guinea-Bissau and The Gambia. If you have any questions about how to interpret these species calls, please get in touch.</p>
</div>
<div class="section" id="snp-sites-and-alleles">
<h2>SNP sites and alleles<a class="headerlink" href="#snp-sites-and-alleles" title="Permalink to this headline">¶</a></h2>
<p>We have called SNP genotypes in all samples at all positions in the genome where the reference allele is not “N”. Data on this set of genomic positions and alleles for a given chromosome arm (e.g., 3R) can be accessed as dask arrays as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pos</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_sites</span><span class="p">(</span><span class="s2">&quot;3R&quot;</span><span class="p">)</span>
<span class="n">pos</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 208.91 MB </td> <td> 134.22 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (33554432,) </td></tr>
    <tr><th> Count </th><td> 3 Tasks </td><td> 2 Chunks </td></tr>
    <tr><th> Type </th><td> int32 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="77" y1="0" x2="77" y2="25" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.000000,0.000000 120.000000,0.000000 120.000000,25.412617 0.000000,25.412617" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ref</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 52.23 MB </td> <td> 52.23 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (52226568,) </td></tr>
    <tr><th> Count </th><td> 2 Tasks </td><td> 1 Chunks </td></tr>
    <tr><th> Type </th><td> |S1 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.000000,0.000000 120.000000,0.000000 120.000000,25.412617 0.000000,25.412617" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">alt</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 156.68 MB </td> <td> 133.69 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 3) </td> <td> (44564480, 3) </td></tr>
    <tr><th> Count </th><td> 3 Tasks </td><td> 2 Chunks </td></tr>
    <tr><th> Type </th><td> |S1 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="75" height="170" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="25" y2="0" style="stroke-width:2" />
  <line x1="0" y1="102" x2="25" y2="102" />
  <line x1="0" y1="120" x2="25" y2="120" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="120" style="stroke-width:2" />
  <line x1="25" y1="0" x2="25" y2="120" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.000000,0.000000 25.412617,0.000000 25.412617,120.000000 0.000000,120.000000" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="12.706308" y="140.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" >3</text>
  <text x="45.412617" y="60.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,45.412617,60.000000)">52226568</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<p>Data can be loaded into memory as numpy arrays as shown in the following examples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read first 10 SNP positions into a numpy array</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10], dtype=int32)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read first 10 SNP reference alleles</span>
<span class="n">r</span> <span class="o">=</span> <span class="n">ref</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="n">r</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([b&#39;C&#39;, b&#39;C&#39;, b&#39;T&#39;, b&#39;C&#39;, b&#39;T&#39;, b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;, b&#39;T&#39;, b&#39;T&#39;],
      dtype=&#39;|S1&#39;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read first 10 SNP alternate alleles</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">alt</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="n">a</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[b&#39;A&#39;, b&#39;T&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;T&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;T&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;],
       [b&#39;C&#39;, b&#39;T&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;T&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;C&#39;, b&#39;T&#39;],
       [b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;]], dtype=&#39;|S1&#39;)
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="site-filters">
<h2>Site filters<a class="headerlink" href="#site-filters" title="Permalink to this headline">¶</a></h2>
<p>SNP calling is not always reliable, and we have created some site filters to allow excluding low quality SNPs.</p>
<p>Because different species may have different genome accessibility issues, we have created three separate site filters:</p>
<ul class="simple">
<li><p>The “gamb_colu” filter is design for working only with samples that <strong>are not</strong> An. arabiensis.</p></li>
<li><p>The “arab” filter is designed for when only working with samples that <strong>are</strong> An. arabiensis.</p></li>
<li><p>The “gamb_colu_arab” filter is suitable for when analysing samples of any species together.</p></li>
</ul>
<p>Each set of site filters provides a “filter_pass” Boolean mask for each chromosome arm, where True indicates that the site passed the filter and is accessible to high quality SNP calling.</p>
<p>The site filters data can be accessed as dask arrays as shown in the examples below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># access gamb_colu_arab site filters for chromosome arm 3R as a dask array</span>
<span class="n">filter_pass</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">site_filters</span><span class="p">(</span><span class="s2">&quot;3R&quot;</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="s2">&quot;gamb_colu_arab&quot;</span><span class="p">)</span>
<span class="n">filter_pass</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 52.23 MB </td> <td> 52.23 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (52226568,) </td></tr>
    <tr><th> Count </th><td> 2 Tasks </td><td> 1 Chunks </td></tr>
    <tr><th> Type </th><td> bool </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.000000,0.000000 120.000000,0.000000 120.000000,25.412617 0.000000,25.412617" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read filter values for first 10 SNPs (True means the site passes filters)</span>
<span class="n">f</span> <span class="o">=</span> <span class="n">filter_pass</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="n">f</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([False, False, False, False, False, False, False, False, False,
       False])
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="snp-genotypes">
<h2>SNP genotypes<a class="headerlink" href="#snp-genotypes" title="Permalink to this headline">¶</a></h2>
<p>SNP genotypes for individual samples are available. Genotypes are stored as a three-dimensional array, where the first dimension corresponds to genomic positions, the second dimension is samples, and the third dimension is ploidy (2). Values coded as integers, where -1 represents a missing value, 0 represents the reference allele, and 1, 2, and 3 represent alternate alleles.</p>
<p>SNP genotypes can be accessed as dask arrays as shown below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gt</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_genotypes</span><span class="p">(</span><span class="s2">&quot;3R&quot;</span><span class="p">,</span> <span class="n">cohort</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AG1000G-BF-A&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-B&quot;</span><span class="p">])</span>
<span class="n">gt</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 29.56 GB </td> <td> 122.40 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 283, 2) </td> <td> (600000, 102, 2) </td></tr>
    <tr><th> Count </th><td> 530 Tasks </td><td> 264 Chunks </td></tr>
    <tr><th> Type </th><td> int8 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="156" height="146" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="10" y1="8" x2="80" y2="79" />
  <line x1="10" y1="16" x2="80" y2="86" />
  <line x1="10" y1="25" x2="80" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="10" y2="25" style="stroke-width:2" />
  <line x1="10" y1="0" x2="10" y2="26" />
  <line x1="11" y1="1" x2="11" y2="27" />
  <line x1="12" y1="2" x2="12" y2="27" />
  <line x1="13" y1="3" x2="13" y2="28" />
  <line x1="14" y1="4" x2="14" y2="29" />
  <line x1="14" y1="4" x2="14" y2="30" />
  <line x1="15" y1="5" x2="15" y2="31" />
  <line x1="16" y1="6" x2="16" y2="31" />
  <line x1="17" y1="7" x2="17" y2="32" />
  <line x1="18" y1="8" x2="18" y2="33" />
  <line x1="18" y1="8" x2="18" y2="34" />
  <line x1="19" y1="9" x2="19" y2="35" />
  <line x1="20" y1="10" x2="20" y2="35" />
  <line x1="21" y1="11" x2="21" y2="36" />
  <line x1="22" y1="12" x2="22" y2="37" />
  <line x1="22" y1="12" x2="22" y2="38" />
  <line x1="23" y1="13" x2="23" y2="39" />
  <line x1="24" y1="14" x2="24" y2="40" />
  <line x1="25" y1="15" x2="25" y2="40" />
  <line x1="26" y1="16" x2="26" y2="41" />
  <line x1="27" y1="17" x2="27" y2="42" />
  <line x1="27" y1="17" x2="27" y2="43" />
  <line x1="28" y1="18" x2="28" y2="44" />
  <line x1="29" y1="19" x2="29" y2="44" />
  <line x1="30" y1="20" x2="30" y2="45" />
  <line x1="31" y1="21" x2="31" y2="46" />
  <line x1="31" y1="21" x2="31" y2="47" />
  <line x1="32" y1="22" x2="32" y2="48" />
  <line x1="33" y1="23" x2="33" y2="48" />
  <line x1="34" y1="24" x2="34" y2="49" />
  <line x1="35" y1="25" x2="35" y2="50" />
  <line x1="35" y1="25" x2="35" y2="51" />
  <line x1="36" y1="26" x2="36" y2="52" />
  <line x1="37" y1="27" x2="37" y2="52" />
  <line x1="38" y1="28" x2="38" y2="53" />
  <line x1="39" y1="29" x2="39" y2="54" />
  <line x1="40" y1="30" x2="40" y2="55" />
  <line x1="40" y1="30" x2="40" y2="56" />
  <line x1="41" y1="31" x2="41" y2="57" />
  <line x1="42" y1="32" x2="42" y2="57" />
  <line x1="43" y1="33" x2="43" y2="58" />
  <line x1="44" y1="34" x2="44" y2="59" />
  <line x1="44" y1="34" x2="44" y2="60" />
  <line x1="45" y1="35" x2="45" y2="61" />
  <line x1="46" y1="36" x2="46" y2="61" />
  <line x1="47" y1="37" x2="47" y2="62" />
  <line x1="48" y1="38" x2="48" y2="63" />
  <line x1="48" y1="38" x2="48" y2="64" />
  <line x1="49" y1="39" x2="49" y2="65" />
  <line x1="50" y1="40" x2="50" y2="65" />
  <line x1="51" y1="41" x2="51" y2="66" />
  <line x1="52" y1="42" x2="52" y2="67" />
  <line x1="52" y1="42" x2="52" y2="68" />
  <line x1="53" y1="43" x2="53" y2="69" />
  <line x1="54" y1="44" x2="54" y2="70" />
  <line x1="55" y1="45" x2="55" y2="70" />
  <line x1="56" y1="46" x2="56" y2="71" />
  <line x1="57" y1="47" x2="57" y2="72" />
  <line x1="57" y1="47" x2="57" y2="73" />
  <line x1="58" y1="48" x2="58" y2="74" />
  <line x1="59" y1="49" x2="59" y2="74" />
  <line x1="60" y1="50" x2="60" y2="75" />
  <line x1="61" y1="51" x2="61" y2="76" />
  <line x1="61" y1="51" x2="61" y2="77" />
  <line x1="62" y1="52" x2="62" y2="78" />
  <line x1="63" y1="53" x2="63" y2="78" />
  <line x1="64" y1="54" x2="64" y2="79" />
  <line x1="65" y1="55" x2="65" y2="80" />
  <line x1="65" y1="55" x2="65" y2="81" />
  <line x1="66" y1="56" x2="66" y2="82" />
  <line x1="67" y1="57" x2="67" y2="82" />
  <line x1="68" y1="58" x2="68" y2="83" />
  <line x1="69" y1="59" x2="69" y2="84" />
  <line x1="70" y1="60" x2="70" y2="85" />
  <line x1="70" y1="60" x2="70" y2="86" />
  <line x1="71" y1="61" x2="71" y2="87" />
  <line x1="72" y1="62" x2="72" y2="87" />
  <line x1="73" y1="63" x2="73" y2="88" />
  <line x1="74" y1="64" x2="74" y2="89" />
  <line x1="74" y1="64" x2="74" y2="90" />
  <line x1="75" y1="65" x2="75" y2="91" />
  <line x1="76" y1="66" x2="76" y2="91" />
  <line x1="77" y1="67" x2="77" y2="92" />
  <line x1="78" y1="68" x2="78" y2="93" />
  <line x1="78" y1="68" x2="78" y2="94" />
  <line x1="79" y1="69" x2="79" y2="95" />
  <line x1="80" y1="70" x2="80" y2="95" />
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.000000,0.000000 80.588235,70.588235 80.588235,96.000852 10.000000,25.412617" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="35" y2="0" style="stroke-width:2" />
  <line x1="10" y1="0" x2="36" y2="0" />
  <line x1="11" y1="1" x2="37" y2="1" />
  <line x1="12" y1="2" x2="37" y2="2" />
  <line x1="13" y1="3" x2="38" y2="3" />
  <line x1="14" y1="4" x2="39" y2="4" />
  <line x1="14" y1="4" x2="40" y2="4" />
  <line x1="15" y1="5" x2="41" y2="5" />
  <line x1="16" y1="6" x2="41" y2="6" />
  <line x1="17" y1="7" x2="42" y2="7" />
  <line x1="18" y1="8" x2="43" y2="8" />
  <line x1="18" y1="8" x2="44" y2="8" />
  <line x1="19" y1="9" x2="45" y2="9" />
  <line x1="20" y1="10" x2="45" y2="10" />
  <line x1="21" y1="11" x2="46" y2="11" />
  <line x1="22" y1="12" x2="47" y2="12" />
  <line x1="22" y1="12" x2="48" y2="12" />
  <line x1="23" y1="13" x2="49" y2="13" />
  <line x1="24" y1="14" x2="50" y2="14" />
  <line x1="25" y1="15" x2="50" y2="15" />
  <line x1="26" y1="16" x2="51" y2="16" />
  <line x1="27" y1="17" x2="52" y2="17" />
  <line x1="27" y1="17" x2="53" y2="17" />
  <line x1="28" y1="18" x2="54" y2="18" />
  <line x1="29" y1="19" x2="54" y2="19" />
  <line x1="30" y1="20" x2="55" y2="20" />
  <line x1="31" y1="21" x2="56" y2="21" />
  <line x1="31" y1="21" x2="57" y2="21" />
  <line x1="32" y1="22" x2="58" y2="22" />
  <line x1="33" y1="23" x2="58" y2="23" />
  <line x1="34" y1="24" x2="59" y2="24" />
  <line x1="35" y1="25" x2="60" y2="25" />
  <line x1="35" y1="25" x2="61" y2="25" />
  <line x1="36" y1="26" x2="62" y2="26" />
  <line x1="37" y1="27" x2="62" y2="27" />
  <line x1="38" y1="28" x2="63" y2="28" />
  <line x1="39" y1="29" x2="64" y2="29" />
  <line x1="40" y1="30" x2="65" y2="30" />
  <line x1="40" y1="30" x2="66" y2="30" />
  <line x1="41" y1="31" x2="67" y2="31" />
  <line x1="42" y1="32" x2="67" y2="32" />
  <line x1="43" y1="33" x2="68" y2="33" />
  <line x1="44" y1="34" x2="69" y2="34" />
  <line x1="44" y1="34" x2="70" y2="34" />
  <line x1="45" y1="35" x2="71" y2="35" />
  <line x1="46" y1="36" x2="71" y2="36" />
  <line x1="47" y1="37" x2="72" y2="37" />
  <line x1="48" y1="38" x2="73" y2="38" />
  <line x1="48" y1="38" x2="74" y2="38" />
  <line x1="49" y1="39" x2="75" y2="39" />
  <line x1="50" y1="40" x2="75" y2="40" />
  <line x1="51" y1="41" x2="76" y2="41" />
  <line x1="52" y1="42" x2="77" y2="42" />
  <line x1="52" y1="42" x2="78" y2="42" />
  <line x1="53" y1="43" x2="79" y2="43" />
  <line x1="54" y1="44" x2="80" y2="44" />
  <line x1="55" y1="45" x2="80" y2="45" />
  <line x1="56" y1="46" x2="81" y2="46" />
  <line x1="57" y1="47" x2="82" y2="47" />
  <line x1="57" y1="47" x2="83" y2="47" />
  <line x1="58" y1="48" x2="84" y2="48" />
  <line x1="59" y1="49" x2="84" y2="49" />
  <line x1="60" y1="50" x2="85" y2="50" />
  <line x1="61" y1="51" x2="86" y2="51" />
  <line x1="61" y1="51" x2="87" y2="51" />
  <line x1="62" y1="52" x2="88" y2="52" />
  <line x1="63" y1="53" x2="88" y2="53" />
  <line x1="64" y1="54" x2="89" y2="54" />
  <line x1="65" y1="55" x2="90" y2="55" />
  <line x1="65" y1="55" x2="91" y2="55" />
  <line x1="66" y1="56" x2="92" y2="56" />
  <line x1="67" y1="57" x2="92" y2="57" />
  <line x1="68" y1="58" x2="93" y2="58" />
  <line x1="69" y1="59" x2="94" y2="59" />
  <line x1="70" y1="60" x2="95" y2="60" />
  <line x1="70" y1="60" x2="96" y2="60" />
  <line x1="71" y1="61" x2="97" y2="61" />
  <line x1="72" y1="62" x2="97" y2="62" />
  <line x1="73" y1="63" x2="98" y2="63" />
  <line x1="74" y1="64" x2="99" y2="64" />
  <line x1="74" y1="64" x2="100" y2="64" />
  <line x1="75" y1="65" x2="101" y2="65" />
  <line x1="76" y1="66" x2="101" y2="66" />
  <line x1="77" y1="67" x2="102" y2="67" />
  <line x1="78" y1="68" x2="103" y2="68" />
  <line x1="78" y1="68" x2="104" y2="68" />
  <line x1="79" y1="69" x2="105" y2="69" />
  <line x1="80" y1="70" x2="105" y2="70" />
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="35" y1="0" x2="106" y2="70" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.000000,0.000000 35.412617,0.000000 106.000852,70.588235 80.588235,70.588235" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />
  <line x1="80" y1="79" x2="106" y2="79" />
  <line x1="80" y1="86" x2="106" y2="86" />
  <line x1="80" y1="96" x2="106" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />
  <line x1="106" y1="70" x2="106" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="80.588235,70.588235 106.000852,70.588235 106.000852,96.000852 80.588235,96.000852" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="93.294544" y="116.000852" font-size="1.0rem" font-weight="100" text-anchor="middle" >2</text>
  <text x="126.000852" y="83.294544" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,126.000852,83.294544)">283</text>
  <text x="35.294118" y="80.706734" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,35.294118,80.706734)">52226568</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<p>Note that the columns of this array (second dimension) match the rows in the sample metadata, if the same sample sets were loaded. I.e.:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seq_id</span> <span class="o">=</span> <span class="s1">&#39;3R&#39;</span>
<span class="n">cohort</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AG1000G-BF-A&#39;</span><span class="p">,</span> <span class="s1">&#39;AG1000G-BF-B&#39;</span><span class="p">]</span>
<span class="n">df_samples</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">(</span><span class="n">cohort</span><span class="o">=</span><span class="n">cohort</span><span class="p">)</span>
<span class="n">gt</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_genotypes</span><span class="p">(</span><span class="n">seq_id</span><span class="o">=</span><span class="n">seq_id</span><span class="p">,</span> <span class="n">cohort</span><span class="o">=</span><span class="n">cohort</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">df_samples</span><span class="p">)</span> <span class="o">==</span> <span class="n">gt</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
<p>You can use this correspondance to apply further subsetting operations to the genotypes by querying the sample metadata. E.g.:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dask.array</span> <span class="k">as</span> <span class="nn">da</span>
<span class="n">loc_gambiae</span> <span class="o">=</span> <span class="n">df_samples</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;species == &#39;gambiae&#39;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">loc_gambiae</span><span class="p">)</span><span class="si">}</span><span class="s2"> gambiae samples&quot;</span><span class="p">)</span>
<span class="n">gt_gambiae</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="n">gt</span><span class="p">,</span> <span class="n">loc_gambiae</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">gt_gambiae</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>found 144 gambiae samples
</pre></div>
</div>
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 15.04 GB </td> <td> 64.80 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 144, 2) </td> <td> (600000, 54, 2) </td></tr>
    <tr><th> Count </th><td> 794 Tasks </td><td> 264 Chunks </td></tr>
    <tr><th> Type </th><td> int8 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="156" height="146" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="10" y1="9" x2="80" y2="80" />
  <line x1="10" y1="17" x2="80" y2="87" />
  <line x1="10" y1="25" x2="80" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="10" y2="25" style="stroke-width:2" />
  <line x1="10" y1="0" x2="10" y2="26" />
  <line x1="11" y1="1" x2="11" y2="27" />
  <line x1="12" y1="2" x2="12" y2="27" />
  <line x1="13" y1="3" x2="13" y2="28" />
  <line x1="14" y1="4" x2="14" y2="29" />
  <line x1="14" y1="4" x2="14" y2="30" />
  <line x1="15" y1="5" x2="15" y2="31" />
  <line x1="16" y1="6" x2="16" y2="31" />
  <line x1="17" y1="7" x2="17" y2="32" />
  <line x1="18" y1="8" x2="18" y2="33" />
  <line x1="18" y1="8" x2="18" y2="34" />
  <line x1="19" y1="9" x2="19" y2="35" />
  <line x1="20" y1="10" x2="20" y2="35" />
  <line x1="21" y1="11" x2="21" y2="36" />
  <line x1="22" y1="12" x2="22" y2="37" />
  <line x1="22" y1="12" x2="22" y2="38" />
  <line x1="23" y1="13" x2="23" y2="39" />
  <line x1="24" y1="14" x2="24" y2="40" />
  <line x1="25" y1="15" x2="25" y2="40" />
  <line x1="26" y1="16" x2="26" y2="41" />
  <line x1="27" y1="17" x2="27" y2="42" />
  <line x1="27" y1="17" x2="27" y2="43" />
  <line x1="28" y1="18" x2="28" y2="44" />
  <line x1="29" y1="19" x2="29" y2="44" />
  <line x1="30" y1="20" x2="30" y2="45" />
  <line x1="31" y1="21" x2="31" y2="46" />
  <line x1="31" y1="21" x2="31" y2="47" />
  <line x1="32" y1="22" x2="32" y2="48" />
  <line x1="33" y1="23" x2="33" y2="48" />
  <line x1="34" y1="24" x2="34" y2="49" />
  <line x1="35" y1="25" x2="35" y2="50" />
  <line x1="35" y1="25" x2="35" y2="51" />
  <line x1="36" y1="26" x2="36" y2="52" />
  <line x1="37" y1="27" x2="37" y2="52" />
  <line x1="38" y1="28" x2="38" y2="53" />
  <line x1="39" y1="29" x2="39" y2="54" />
  <line x1="40" y1="30" x2="40" y2="55" />
  <line x1="40" y1="30" x2="40" y2="56" />
  <line x1="41" y1="31" x2="41" y2="57" />
  <line x1="42" y1="32" x2="42" y2="57" />
  <line x1="43" y1="33" x2="43" y2="58" />
  <line x1="44" y1="34" x2="44" y2="59" />
  <line x1="44" y1="34" x2="44" y2="60" />
  <line x1="45" y1="35" x2="45" y2="61" />
  <line x1="46" y1="36" x2="46" y2="61" />
  <line x1="47" y1="37" x2="47" y2="62" />
  <line x1="48" y1="38" x2="48" y2="63" />
  <line x1="48" y1="38" x2="48" y2="64" />
  <line x1="49" y1="39" x2="49" y2="65" />
  <line x1="50" y1="40" x2="50" y2="65" />
  <line x1="51" y1="41" x2="51" y2="66" />
  <line x1="52" y1="42" x2="52" y2="67" />
  <line x1="52" y1="42" x2="52" y2="68" />
  <line x1="53" y1="43" x2="53" y2="69" />
  <line x1="54" y1="44" x2="54" y2="70" />
  <line x1="55" y1="45" x2="55" y2="70" />
  <line x1="56" y1="46" x2="56" y2="71" />
  <line x1="57" y1="47" x2="57" y2="72" />
  <line x1="57" y1="47" x2="57" y2="73" />
  <line x1="58" y1="48" x2="58" y2="74" />
  <line x1="59" y1="49" x2="59" y2="74" />
  <line x1="60" y1="50" x2="60" y2="75" />
  <line x1="61" y1="51" x2="61" y2="76" />
  <line x1="61" y1="51" x2="61" y2="77" />
  <line x1="62" y1="52" x2="62" y2="78" />
  <line x1="63" y1="53" x2="63" y2="78" />
  <line x1="64" y1="54" x2="64" y2="79" />
  <line x1="65" y1="55" x2="65" y2="80" />
  <line x1="65" y1="55" x2="65" y2="81" />
  <line x1="66" y1="56" x2="66" y2="82" />
  <line x1="67" y1="57" x2="67" y2="82" />
  <line x1="68" y1="58" x2="68" y2="83" />
  <line x1="69" y1="59" x2="69" y2="84" />
  <line x1="70" y1="60" x2="70" y2="85" />
  <line x1="70" y1="60" x2="70" y2="86" />
  <line x1="71" y1="61" x2="71" y2="87" />
  <line x1="72" y1="62" x2="72" y2="87" />
  <line x1="73" y1="63" x2="73" y2="88" />
  <line x1="74" y1="64" x2="74" y2="89" />
  <line x1="74" y1="64" x2="74" y2="90" />
  <line x1="75" y1="65" x2="75" y2="91" />
  <line x1="76" y1="66" x2="76" y2="91" />
  <line x1="77" y1="67" x2="77" y2="92" />
  <line x1="78" y1="68" x2="78" y2="93" />
  <line x1="78" y1="68" x2="78" y2="94" />
  <line x1="79" y1="69" x2="79" y2="95" />
  <line x1="80" y1="70" x2="80" y2="95" />
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.000000,0.000000 80.588235,70.588235 80.588235,96.000852 10.000000,25.412617" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="35" y2="0" style="stroke-width:2" />
  <line x1="10" y1="0" x2="36" y2="0" />
  <line x1="11" y1="1" x2="37" y2="1" />
  <line x1="12" y1="2" x2="37" y2="2" />
  <line x1="13" y1="3" x2="38" y2="3" />
  <line x1="14" y1="4" x2="39" y2="4" />
  <line x1="14" y1="4" x2="40" y2="4" />
  <line x1="15" y1="5" x2="41" y2="5" />
  <line x1="16" y1="6" x2="41" y2="6" />
  <line x1="17" y1="7" x2="42" y2="7" />
  <line x1="18" y1="8" x2="43" y2="8" />
  <line x1="18" y1="8" x2="44" y2="8" />
  <line x1="19" y1="9" x2="45" y2="9" />
  <line x1="20" y1="10" x2="45" y2="10" />
  <line x1="21" y1="11" x2="46" y2="11" />
  <line x1="22" y1="12" x2="47" y2="12" />
  <line x1="22" y1="12" x2="48" y2="12" />
  <line x1="23" y1="13" x2="49" y2="13" />
  <line x1="24" y1="14" x2="50" y2="14" />
  <line x1="25" y1="15" x2="50" y2="15" />
  <line x1="26" y1="16" x2="51" y2="16" />
  <line x1="27" y1="17" x2="52" y2="17" />
  <line x1="27" y1="17" x2="53" y2="17" />
  <line x1="28" y1="18" x2="54" y2="18" />
  <line x1="29" y1="19" x2="54" y2="19" />
  <line x1="30" y1="20" x2="55" y2="20" />
  <line x1="31" y1="21" x2="56" y2="21" />
  <line x1="31" y1="21" x2="57" y2="21" />
  <line x1="32" y1="22" x2="58" y2="22" />
  <line x1="33" y1="23" x2="58" y2="23" />
  <line x1="34" y1="24" x2="59" y2="24" />
  <line x1="35" y1="25" x2="60" y2="25" />
  <line x1="35" y1="25" x2="61" y2="25" />
  <line x1="36" y1="26" x2="62" y2="26" />
  <line x1="37" y1="27" x2="62" y2="27" />
  <line x1="38" y1="28" x2="63" y2="28" />
  <line x1="39" y1="29" x2="64" y2="29" />
  <line x1="40" y1="30" x2="65" y2="30" />
  <line x1="40" y1="30" x2="66" y2="30" />
  <line x1="41" y1="31" x2="67" y2="31" />
  <line x1="42" y1="32" x2="67" y2="32" />
  <line x1="43" y1="33" x2="68" y2="33" />
  <line x1="44" y1="34" x2="69" y2="34" />
  <line x1="44" y1="34" x2="70" y2="34" />
  <line x1="45" y1="35" x2="71" y2="35" />
  <line x1="46" y1="36" x2="71" y2="36" />
  <line x1="47" y1="37" x2="72" y2="37" />
  <line x1="48" y1="38" x2="73" y2="38" />
  <line x1="48" y1="38" x2="74" y2="38" />
  <line x1="49" y1="39" x2="75" y2="39" />
  <line x1="50" y1="40" x2="75" y2="40" />
  <line x1="51" y1="41" x2="76" y2="41" />
  <line x1="52" y1="42" x2="77" y2="42" />
  <line x1="52" y1="42" x2="78" y2="42" />
  <line x1="53" y1="43" x2="79" y2="43" />
  <line x1="54" y1="44" x2="80" y2="44" />
  <line x1="55" y1="45" x2="80" y2="45" />
  <line x1="56" y1="46" x2="81" y2="46" />
  <line x1="57" y1="47" x2="82" y2="47" />
  <line x1="57" y1="47" x2="83" y2="47" />
  <line x1="58" y1="48" x2="84" y2="48" />
  <line x1="59" y1="49" x2="84" y2="49" />
  <line x1="60" y1="50" x2="85" y2="50" />
  <line x1="61" y1="51" x2="86" y2="51" />
  <line x1="61" y1="51" x2="87" y2="51" />
  <line x1="62" y1="52" x2="88" y2="52" />
  <line x1="63" y1="53" x2="88" y2="53" />
  <line x1="64" y1="54" x2="89" y2="54" />
  <line x1="65" y1="55" x2="90" y2="55" />
  <line x1="65" y1="55" x2="91" y2="55" />
  <line x1="66" y1="56" x2="92" y2="56" />
  <line x1="67" y1="57" x2="92" y2="57" />
  <line x1="68" y1="58" x2="93" y2="58" />
  <line x1="69" y1="59" x2="94" y2="59" />
  <line x1="70" y1="60" x2="95" y2="60" />
  <line x1="70" y1="60" x2="96" y2="60" />
  <line x1="71" y1="61" x2="97" y2="61" />
  <line x1="72" y1="62" x2="97" y2="62" />
  <line x1="73" y1="63" x2="98" y2="63" />
  <line x1="74" y1="64" x2="99" y2="64" />
  <line x1="74" y1="64" x2="100" y2="64" />
  <line x1="75" y1="65" x2="101" y2="65" />
  <line x1="76" y1="66" x2="101" y2="66" />
  <line x1="77" y1="67" x2="102" y2="67" />
  <line x1="78" y1="68" x2="103" y2="68" />
  <line x1="78" y1="68" x2="104" y2="68" />
  <line x1="79" y1="69" x2="105" y2="69" />
  <line x1="80" y1="70" x2="105" y2="70" />
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="35" y1="0" x2="106" y2="70" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.000000,0.000000 35.412617,0.000000 106.000852,70.588235 80.588235,70.588235" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />
  <line x1="80" y1="80" x2="106" y2="80" />
  <line x1="80" y1="87" x2="106" y2="87" />
  <line x1="80" y1="96" x2="106" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />
  <line x1="106" y1="70" x2="106" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="80.588235,70.588235 106.000852,70.588235 106.000852,96.000852 80.588235,96.000852" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="93.294544" y="116.000852" font-size="1.0rem" font-weight="100" text-anchor="middle" >2</text>
  <text x="126.000852" y="83.294544" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,126.000852,83.294544)">144</text>
  <text x="35.294118" y="80.706734" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,35.294118,80.706734)">52226568</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<p>Data can be read into memory as numpy arrays, e.g., read genotypes for the first 5 SNPs and the first 3 samples:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">g</span> <span class="o">=</span> <span class="n">gt</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="n">g</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[[0, 0],
        [0, 0],
        [0, 0]],

       [[0, 0],
        [0, 0],
        [0, 0]],

       [[0, 0],
        [0, 0],
        [0, 0]],

       [[0, 0],
        [0, 0],
        [0, 0]],

       [[0, 0],
        [0, 0],
        [0, 0]]], dtype=int8)
</pre></div>
</div>
</div>
</div>
<p>If you want to work with the genotype calls, you may find it convenient to use <a class="reference external" href="http://scikit-allel.readthedocs.org/">scikit-allel</a>. E.g., this code sets up a genotype array:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># use the scikit-allel wrapper class for genotype calls</span>
<span class="kn">import</span> <span class="nn">allel</span>
<span class="n">gt</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_genotypes</span><span class="p">(</span><span class="s2">&quot;3R&quot;</span><span class="p">,</span> <span class="n">cohort</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AG1000G-BF-A&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-B&quot;</span><span class="p">])</span>
<span class="n">gt</span> <span class="o">=</span> <span class="n">allel</span><span class="o">.</span><span class="n">GenotypeDaskArray</span><span class="p">(</span><span class="n">gt</span><span class="p">)</span>
<span class="n">gt</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="allel allel-DisplayAs2D"><span>&lt;GenotypeDaskArray shape=(52226568, 283, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style="text-align: center">0</th><th style="text-align: center">1</th><th style="text-align: center">2</th><th style="text-align: center">3</th><th style="text-align: center">4</th><th style="text-align: center">...</th><th style="text-align: center">278</th><th style="text-align: center">279</th><th style="text-align: center">280</th><th style="text-align: center">281</th><th style="text-align: center">282</th></tr></thead><tbody><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">0</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">1</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">2</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">...</th><td style="text-align: center" colspan="12">...</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">52226565</th><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">52226566</th><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">...</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">52226567</th><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">...</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td></tr></tbody></table></div></div></div>
</div>
<p>Here’s an example computation to count the number of segregating SNPs on chromosome arm 3R that also pass gamb_colu_arab site filters. This will take a little while, because it is scanning millions of genotype calls in hundreds of samples:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">allel</span>

<span class="c1"># import dask progress bar</span>
<span class="kn">from</span> <span class="nn">dask.diagnostics.progress</span> <span class="kn">import</span> <span class="n">ProgressBar</span>

<span class="c1"># import numpy</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="c1"># choose chromosome arm</span>
<span class="n">seq_id</span> <span class="o">=</span> <span class="s2">&quot;3R&quot;</span>

<span class="c1"># choose site filter mask</span>
<span class="n">mask</span> <span class="o">=</span> <span class="s2">&quot;gamb_colu_arab&quot;</span>

<span class="c1"># choose cohort</span>
<span class="n">cohort</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AG1000G-BF-A&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-B&quot;</span><span class="p">]</span>

<span class="c1"># locate pass sites</span>
<span class="n">loc_pass</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">site_filters</span><span class="p">(</span><span class="n">seq_id</span><span class="o">=</span><span class="n">seq_id</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">)</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>

<span class="c1"># perform an allele count over genotypes</span>
<span class="n">gt</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_genotypes</span><span class="p">(</span><span class="n">seq_id</span><span class="o">=</span><span class="n">seq_id</span><span class="p">,</span> <span class="n">cohort</span><span class="o">=</span><span class="n">cohort</span><span class="p">)</span>
<span class="n">gt</span> <span class="o">=</span> <span class="n">allel</span><span class="o">.</span><span class="n">GenotypeDaskArray</span><span class="p">(</span><span class="n">gt</span><span class="p">)</span>
<span class="k">with</span> <span class="n">ProgressBar</span><span class="p">():</span>
    <span class="n">ac</span> <span class="o">=</span> <span class="n">gt</span><span class="o">.</span><span class="n">count_alleles</span><span class="p">(</span><span class="n">max_allele</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
    
<span class="c1"># locate segregating sites</span>
<span class="n">loc_seg</span> <span class="o">=</span> <span class="n">ac</span><span class="o">.</span><span class="n">is_segregating</span><span class="p">()</span>

<span class="c1"># count segregating and pass sites</span>
<span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">loc_pass</span> <span class="o">&amp;</span> <span class="n">loc_seg</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[########################################] | 100% Completed |  2min 32.2s
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>11226723
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="further-information">
<h2>Further information<a class="headerlink" href="#further-information" title="Permalink to this headline">¶</a></h2>
<p>For further information about the Ag1000G project, please visit the <a class="reference external" href="https://www.malariagen.net/ag1000g">Ag1000G website</a>.</p>
<p>For further examples of analysing data please follow <a class="reference external" href="https://twitter.com/malariagenomics">&#64;malariagenomics</a> on Twitter or keep an eye on the <a class="reference external" href="https://github.com/malariagen/vector-public-data/discussions">malariagen/vector-public-data GitHub discussion board</a>.</p>
<p>You may also be interested in the following related documents:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://storage.googleapis.com/vo_agam_release/v3/ag1000g-phase3-contributing-studies.pdf">Ag1000G phase 3 contributing studies</a> - Provides information about the studies that collected and contributed specimens to Ag1000G phase 3, including collection information, related publications and key people involved.</p></li>
<li><p><a class="reference external" href="https://storage.googleapis.com/vo_agam_release/v3/ag1000g-phase3-snp-calling-methods.pdf">Ag1000G phase 3 SNP calling methods</a> - Provides information about the methods used for sequencing, sequence read alignment, SNP calling, and quality control.</p></li>
<li><p><span class="xref myst">Ag1000G phase 3 SNP data release - data download guide &#64;&#64;TODO</span> - A guide to downloading data from this data release and accessing them from a local file system.</p></li>
</ul>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./ag3-snp"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="intro.html" title="previous page">Ag1000G phase 3 - summary of samples</a>
    <a class='right-next' id="next-link" href="download.html" title="next page">Ag1000G phase 3 SNP data release - data download guide</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By The Jupyter Book community<br/>
        
            &copy; Copyright 2020.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../_static/js/index.3da636dd464baa7582d2.js"></script>


    
  </body>
</html>